File: src/crewlet/api/routes/integrations.py
"""GET /integrations — the external surfaces, and what has come through them.

Integrations had close to no surface at all. The dashboard branded an
event once it had already been accepted and routed, which means every
failure mode an operator actually hits was invisible: a Mattermost
SiteURL that blinds every browser while agents keep working, a revoked
bot token, a mis-pasted webhook secret. Rejected deliveries are
deliberately never written to the event store — verification runs before
the row is logged, which is correct — so a signature mismatch left no
trace anywhere except the provider's own delivery UI.

Three things are answerable today and are answered here:

* **how each integration is wired** — from the active config revision,
  including whether a signing secret is present (the boolean, never the
  value);
* **what has arrived** — per-source counts and a last-seen timestamp,
  grouped by the database rather than tallied in a browser that holds
  only the last 400 events;
* **what the routing did with it** — the funnel of ``external_notification``
  against ``notification_skipped`` and ``notifications_coalesced``.

One thing is deliberately NOT inferred: health. An idle Slack and a
401-ing Slack look identical in the event store, so silence is reported
as "no traffic seen", never as "healthy" and never as "down". That is
the same three-valued discipline the health surface keeps for engine
facts.
"""

from __future__ import annotations

from typing import Any

from starlette.requests import Request
from starlette.responses import JSONResponse

from crewlet._logging import get_logger

logger = get_logger("api.routes.integrations")

# The surfaces the engine can receive from, and how each one arrives.
# `secret` names the `app.state` attribute holding its verification
# secret, where one exists.
INBOUND: dict[str, dict[str, Any]] = {
    "slack": {
        "path": "/webhooks/slack/{handle}",
        "kind": "webhook",
        "secret": "slack_signing_secrets",
        "per_seat_secret": True,
    },
    "mattermost": {
        # No usable inbound webhook: outgoing webhooks fire only in
        # public channels and carry no thread id, channel type or
        # mentions, so the engine holds one websocket per seat instead.
        "path": "",
        "kind": "websocket",
        "secret": "",
    },
    "jira": {
        "path": "/webhooks/jira",
        "kind": "webhook",
        "secret": "jira_webhook_secret",
    },
    "confluence": {
        "path": "/webhooks/confluence",
        "kind": "webhook",
        "secret": "confluence_webhook_secret",
    },
    "github": {
        "path": "/webhooks/github",
        "kind": "webhook",
        "secret": "github_webhook_secret",
    },
    "gitlab": {
        "path": "/webhooks/gitlab",
        "kind": "webhook",
        "secret": "gitlab_signing_secret",
    },
    "plane": {
        "path": "/webhooks/plane",
        "kind": "webhook",
        "secret": "plane_webhook_secret",
    },
}

# How long a window the traffic figures cover.
TRAFFIC_WINDOW_HOURS = 24

# The routing outcomes, as event types.
_ROUTED = "external_notification"
_SKIPPED = "notification_skipped"
_COALESCED = "notifications_coalesced"


def _iso(value: Any) -> str:
    try:
        return value.isoformat() if value is not None else ""
    except Exception:
        return str(value or "")


def _secret_present(app: Any, spec: dict[str, Any]) -> bool | None:
    """Whether a verification secret is configured.

    Three-valued on purpose: ``None`` means this surface does not use
    one (Mattermost authenticates its websocket with the bot's own
    token), which is a different statement from "configured without a
    secret" — and the latter is a real, silent failure, because the
    webhook route answers 503 to every delivery until one is set.
    """
    attr = spec.get("secret") or ""
    if not attr:
        return None
    value = getattr(app.state, attr, None)
    if spec.get("per_seat_secret"):
        return bool(value)
    return bool(value)


def _configured(app: Any) -> dict[str, dict[str, Any]]:
    """Which integrations the active revision actually configures.

    Read from the redacted org payload the API already holds, so this
    cannot show a credential and cannot disagree with the Configuration
    room about what is set up.
    """
    integrations = getattr(app.state, "integrations_data", None) or {}
    return {key: block for key, block in integrations.items() if key in INBOUND}


def _seat_identities(app: Any) -> dict[str, list[str]]:
    """Per integration, the seats carrying credentials for it.

    Derived from the same org projection everything else reads, so a
    seat's handle here is the handle routing will look for.
    """
    org = getattr(app.state, "org_data", None) or {}
    found: dict[str, list[str]] = {key: [] for key in INBOUND}

    def visit(role: dict[str, Any]) -> None:
        handle = str(role.get("handle") or role.get("name") or "")
        if not handle:
            return
        for key in INBOUND:
            block = (role.get("integrations") or {}).get(key)
            env = (role.get("mcp_env") or {}).get(key)
            contact = role.get("contact") or {}
            has_contact = any(
                str(field).startswith(key) or key in str(field)
                for field in contact
                if contact.get(field)
            )
            if block or env or has_contact:
                found[key].append(handle)

    def walk_unit(unit: dict[str, Any]) -> None:
        for role in unit.get("roles") or []:
            visit(role)
        for child in unit.get("children") or []:
            walk_unit(child)

    for role in org.get("roles") or []:
        visit(role)
    for unit in org.get("units") or []:
        walk_unit(unit)
    return found


async def integrations_payload(app: Any) -> dict[str, Any]:
    """Every configured surface, how it is wired, and its recent traffic."""
    configured = _configured(app)
    identities = _seat_identities(app)

    counts: list[dict[str, Any]] = []
    store = getattr(app.state, "event_store", None)
    counter = getattr(store, "count_events_by_source", None) if store else None
    have_traffic = False
    if counter is not None:
        try:
            counts = await counter(since_hours=TRAFFIC_WINDOW_HOURS)
            have_traffic = True
        except Exception:
            logger.exception("integration_traffic_read_failed")

    # `source` is the bare integration key on a webhook row and
    # `notification_service.<key>` on a routing one.
    def _key_of(source: str) -> str:
        prefix = "notification_service."
        return source[len(prefix) :] if source.startswith(prefix) else source

    per_key: dict[str, dict[str, Any]] = {}
    for row in counts:
        key = _key_of(str(row.get("source") or ""))
        if key not in INBOUND:
            continue
        bucket = per_key.setdefault(
            key,
            {"inbound": 0, "routed": 0, "skipped": 0, "coalesced": 0, "last_at": ""},
        )
        event_type = str(row.get("event_type") or "")
        n = int(row.get("count") or 0)
        if str(row.get("category") or "") == "webhook":
            bucket["inbound"] += n
        if event_type == _ROUTED:
            bucket["routed"] += n
        elif event_type == _SKIPPED:
            bucket["skipped"] += n
        elif event_type == _COALESCED:
            bucket["coalesced"] += n
        last = _iso(row.get("last_at"))
        if last > bucket["last_at"]:
            bucket["last_at"] = last

    rows = []
    for key, spec in INBOUND.items():
        config = configured.get(key)
        traffic = per_key.get(key) or {}
        rows.append(
            {
                "key": key,
                "configured": config is not None,
                "enabled": bool(config and config["enabled"]),
                "url": (config or {}).get("url", ""),
                "workspace": (config or {}).get("workspace", ""),
                "inbound_kind": spec["kind"],
                "inbound_path": spec["path"],
                "secret_present": _secret_present(app, spec),
                "seats": sorted(identities.get(key) or []),
                "inbound": traffic.get("inbound", 0),
                "routed": traffic.get("routed", 0),
                "skipped": traffic.get("skipped", 0),
                "coalesced": traffic.get("coalesced", 0),
                "last_at": traffic.get("last_at", ""),
            }
        )

    return {
        # False means the counts below are not evidence about anything —
        # there is no event store to count over.
        "traffic_known": have_traffic,
        "window_hours": TRAFFIC_WINDOW_HOURS,
        "integrations": rows,
    }


async def get_integrations(request: Request) -> JSONResponse:
    """GET /integrations — the REST twin of the ``integrations`` query."""
    return JSONResponse(await integrations_payload(request.app))

File: static/dashboard/js/views/integrations.js
// Integrations: how work reaches this company, and what happened to it.
//
// The dashboard branded an event once it had already been accepted and
// routed, which meant every failure mode an operator actually hits was
// invisible. Rejected deliveries are never written to the event store —
// verification runs before the row is logged, which is correct — so a
// mis-pasted signing secret left no trace anywhere but the provider's
// own delivery UI, and the engine looked idle.
//
// Two rules this room keeps, because they are the difference between a
// useful panel and a confident one:
//
//   Silence is not health. An idle Slack and a 401-ing Slack are
//   indistinguishable in the event store, so no traffic is reported as
//   "no traffic seen" — never as "down", and never as "healthy".
//
//   A missing secret IS a finding. A webhook route with no secret
//   configured answers 503 to every delivery, which the sender sees as
//   a retry and the operator sees as nothing at all.

import { esc, escAttr, fmtNum, relTime } from "../format.js";
import { icon } from "../icons.js";
import { integrationMeta } from "../state.js";
import { empty, emptyOrPending, sectionHead, skeletonCards } from "../ui.js";

const POLL_MS = 60_000;

export function createIntegrationsView({ query, refresh }) {
  let data = null;
  let failed = false;
  let timer = null;
  let disposed = false;

  async function load() {
    try {
      data = await query("integrations");
      failed = false;
    } catch {
      failed = true;
    }
    if (!disposed) refresh();
  }

  // What the traffic figures are allowed to claim.
  //
  // `since` rather than a fixed window: the count is over the most recent
  // page of deliveries, not the last N hours, so naming a window would be a
  // number the server never measured.
  // What BECAME of the deliveries, beside how many arrived.
  //
  // "128 inbound" alone cannot tell a working integration from one whose
  // every delivery reaches nobody. The two outcome counts are three-valued —
  // null means this node could not read its event log — so an absent one is
  // omitted rather than rendered as a zero the operator would act on. Both
  // zero renders nothing either: a line saying "0 dropped, 0 merged" on every
  // healthy integration is noise that trains a reader to skip the whole line.
  function outcomes(row) {
    const parts = [];
    if (typeof row.skipped === "number" && row.skipped > 0) {
      parts.push(`${esc(fmtNum(row.skipped))} dropped`);
    }
    if (typeof row.coalesced === "number" && row.coalesced > 0) {
      parts.push(`${esc(fmtNum(row.coalesced))} merged`);
    }
    return parts.length ? ` (${parts.join(", ")})` : "";
  }

  function trafficLine(row, known, since) {
    if (!known) {
      return `<span class="int-quiet">no event store — traffic cannot be counted</span>`;
    }
    if (row.last_at) {
      const span = since ? ` since ${esc(relTime(since))}` : "";
      return `<span class="int-live">${esc(fmtNum(row.inbound))} inbound${span}${outcomes(row)} · last ${esc(relTime(row.last_at))}</span>`;
    }
    return `<span class="int-quiet">nothing has arrived — which is not the same as broken</span>`;
  }

  // Whether a delivery from this surface can WAKE A SEAT.
  //
  // Verifying and storing a delivery is the first half; a parser turning it
  // into a notification is the second, and a vendor can have one without the
  // other. Those render identically everywhere else — configured, secret
  // present, deliveries arriving — so this is the only place an operator can
  // see that a surface is ingesting into a void.
  //
  // Null means the server could not say (a standalone API has no engine to
  // ask), which must not render as "no".
  function routingRow(row) {
    if (row.routes === null || row.routes === undefined) {
      return `<div class="int-kv"><span>Routing</span><span class="zero">not visible from this process</span></div>`;
    }
    return `<div class="int-kv"><span>Routing</span>${
      row.routes
        ? `<span>wakes the seats it concerns</span>`
        : `<span class="warn-ink">ingest only — deliveries are verified and stored, and wake nobody</span>`
    }</div>`;
  }

  // Whether a delivery from this surface can be VERIFIED.
  //
  // Two facts, and the gap between them is the failure worth showing. A
  // secret lives in the config as a ${VAR}: `secret_present` says an
  // operator wrote one down, `secret_usable` says this process resolved it
  // to something a route can check a signature with. An unset variable is
  // present-and-unusable — and from everywhere else it looks fine: the
  // config shows a secret, the vendor's settings page shows a healthy hook,
  // and every delivery is refused with nothing naming the variable.
  //
  // `secret_usable` null means the server could not say (a standalone API
  // has no engine to ask), which must not render as "no".
  function secretRow(row) {
    if (row.secret_present === null || row.secret_present === undefined) {
      // A surface with no shared secret is not a surface missing one.
      return `<div class="int-kv"><span>Verification</span><span>the seat's own token</span></div>`;
    }
    if (!row.secret_present) {
      return `<div class="int-kv"><span>Signing secret</span><span class="warn-ink">missing — every delivery is being turned away with a 503</span></div>`;
    }
    if (row.secret_usable === false) {
      return `<div class="int-kv"><span>Signing secret</span><span class="warn-ink">configured, but it did not resolve to a usable key — every delivery is being turned away with a 503</span></div>`;
    }
    return `<div class="int-kv"><span>Signing secret</span>${
      row.secret_usable
        ? `<span>configured and resolved</span>`
        : `<span>configured<span class="zero"> — whether it resolved is not visible from this process</span></span>`
    }</div>`;
  }

  function card(row, known, since) {
    const meta = integrationMeta(row.key);
    return `
      <div class="int-card ${row.configured ? "" : "is-off"}" data-k="int:${escAttr(row.key)}">
        <div class="int-head">
          <span class="int-badge" style="--int-color:${meta.color}">
            ${icon(meta.icon, "sm")}${esc(meta.label)}
          </span>
          ${
            row.configured
              ? row.enabled
                ? ""
                : `<span class="badge">disabled</span>`
              : `<span class="badge">not configured</span>`
          }
          <span style="flex:1"></span>
          ${row.configured ? trafficLine(row, known, since) : ""}
        </div>
        ${
          row.configured
            ? `<div class="int-body">
                 <div class="int-kv"><span>Inbound</span><span class="mono">${
                   row.inbound_kind === "websocket"
                     ? "one websocket per seat — no public URL needed"
                     : esc(row.inbound_path)
                 }</span></div>
                 ${secretRow(row)}
                 ${routingRow(row)}
                 ${row.url ? `<div class="int-kv"><span>Host</span><span class="mono">${esc(row.url)}</span></div>` : ""}
                 ${row.workspace ? `<div class="int-kv"><span>Workspace</span><span class="mono">${esc(row.workspace)}</span></div>` : ""}
                 <div class="int-kv"><span>Seats</span><span>${
                   row.seats.length
                     ? row.seats
                         .map(
                           (h) =>
                             `<span class="chip clickable" data-action="seat" data-seat="${escAttr(h)}">${esc(h)}</span>`,
                         )
                         .join(" ")
                     : `<span class="zero">none carry credentials for this surface</span>`
                 }</span></div>
               </div>`
            : `<div class="int-body int-off">Nothing in the active configuration wires this surface up.</div>`
        }
      </div>`;
  }

  return {
    slices: ["health", "connected", "org"],

    mount() {
      load();
      timer = setInterval(load, POLL_MS);
    },

    destroy() {
      disposed = true;
      if (timer) clearInterval(timer);
      timer = null;
    },

    render(state) {
      if (data === null) {
        return failed
          ? empty(
              "globe",
              "Could not read integrations",
              "The query failed — retrying every minute.",
            )
          : skeletonCards(3);
      }
      const rows = data.integrations || [];
      const configured = rows.filter((r) => r.configured);
      const rest = rows.filter((r) => !r.configured);
      if (!configured.length) {
        return emptyOrPending(
          state,
          skeletonCards(3),
          "globe",
          "No integrations configured",
          "Agents reach the outside world through chat, a tracker, a knowledge base and a code host. Wire one up in the company configuration.",
        );
      }
      return `
        ${sectionHead("globe", "Connected", configured.length)}
        <div class="int-grid">${configured
          .map((r) => card(r, data.traffic_known, data.traffic_since))
          .join("")}</div>
        ${
          rest.length
            ? sectionHead("plug", "Available", rest.length) +
              `<div class="int-grid">${rest
                .map((r) => card(r, data.traffic_known, data.traffic_since))
                .join("")}</div>`
            : ""
        }`;
    },

    __loadForTest: load,
  };
}
